Queue and fire API events for Web Extensions.

kiaraarose opened the pull request:
#### 0ea5fe052ea303c25413797a41e1523d3c3d30c8
<pre>
Queue and fire API events for Web Extensions.
<a href="https://bugs.webkit.org/show_bug.cgi?id=249254">https://bugs.webkit.org/show_bug.cgi?id=249254</a>

Reviewed by Timothy Hatcher.

This patch adds support for queuing and firing API events for web extensions.

Testing:
I added an event listener in the background page of a test extension, and with this code below,
I tested that the completion handler was successfully called.
EventListenerTypeSet set;
set.add(WebExtensionEventListenerType::ActionOnClicked);
fireEvents(set, makeBlockPtr([ protectedThis = Ref { *this }]() {
    RELEASE_LOG(Extensions, &quot;KIARA&quot;);
}));

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::loadBackgroundContentIfNecessary):
(WebKit::WebExtensionContext::scheduleBackgroundContentToUnload):

(WebKit::WebExtensionContext::queueStartupAndInstallEventsForExtensionIfNecessary):
Queue and fire event listeners.

(WebKit::WebExtensionContext::performTasksAfterBackgroundContentLoads):
(WebKit::WebExtensionContext::fireEvents):
Call the completion handler for persistent background pages. For non-persistent backgrounds, wake
up the background page and queue the listener to fire after the background is loaded.

(WebKit::WebExtensionContext::queueEventToFireAfterBackgroundContentLoads):
Add completion handler to queue of listeners to fires.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
(WebKit::WebExtensionAPIEvent::addListener):
(WebKit::WebExtensionAPIEvent::removeListener):
Add the extension context identifier. Without this, we won&apos;t know which context this message is for.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm:
(WebKit::WebExtensionAPIWebNavigationEvent::addListener):
(WebKit::WebExtensionAPIWebNavigationEvent::removeListener):
Add the extension context identifier. Without this, we won&apos;t know which context this message is for.

Canonical link: <a href="https://commits.webkit.org/257958@main">https://commits.webkit.org/257958@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5cd8db3a7264c2b03903594c8b8f8b21f195e28

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100501 "Failed to checkout and rebase branch from PR 7581") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/9650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/33552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/109821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/10578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/107675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106280 "Failed to checkout and rebase branch from PR 7581") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/10578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/33552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/92909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/10578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/33552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/3381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/33552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/3379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/9498 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/33552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/5188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1213,6 +1213,63 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 
+void WebExtensionContext::loadBackgroundPageIfNecessary(CompletionHandler<void()>&& completionHandler)
+{
+    if (!extension().backgroundContentPath() || extension().backgroundContentIsPersistent()) {
```
So this isn't called on start-up? The method name sounds a bit like something that would be called at startup as well as when needed.

kiaraarose replied:
I believe it's only called when we need to perform some tasks but the background page might be asleep so we wake it up. It's currently not being used yet but it seems like something we'll need for `runtime.getBackgroundPage()` and `runtime.sendMessage()`.

kiaraarose replied:
I can change it to `wakeUpBackgroundContentIfNecessary()`?

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -96,6 +97,8 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
 
     using EventListenterTypeCountedSet = HashCountedSet<WebExtensionEventListenerType, WTF::IntHash<WebKit::WebExtensionEventListenerType>, WTF::StrongEnumHashTraits<WebKit::WebExtensionEventListenerType>>;
+    using EventListenerTypeSet = HashSet<WebExtensionEventListenerType, WTF::IntHash<WebKit::WebExtensionEventListenerType>, WTF::StrongEnumHashTraits<WebKit::WebExtensionEventListenerType>>;
+    using ActionsToPerformVector = Vector<CompletionHandler<void()>>;
```
`VoidCompletionHandlerVector` would be a better name, so it can be reused if needed.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1252,9 +1309,54 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+        return;
+    }
+
+    bool backgroundPageIsLoading = m_actionsToPerformAfterBackgroundPageLoads.size();
```
`bool backgroundPageIsLoading = !m_actionsToPerformAfterBackgroundPageLoads.isEmpty();`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1252,9 +1309,54 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    }
+}
+
+void WebExtensionContext::queueEventToFireAfterBackgroundPageLoads(CompletionHandler<void()> &&completionHandler)
```
I'm trying to avoid the term "BackgroundPage" since it can be a service worker now. This comment applies to all the instances in this PR.

`queueEventToFireAfterBackgroundContentLoads`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1213,6 +1213,63 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+        [m_state removeObjectForKey:backgroundPageListenersStateKey];
+
+        // FIXME: <https://webkit.org/b/248889> Set queued install event details.
+    } else if (!didQueueStartupEvent)
```
This needs `{ }` since the comment isn't enough, I don't think.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -0,0 +1,44 @@
+namespace WebKit {
+}
+
+#endif // ENABLE(WK_WEB_EXTENSIONS)
```
Did you mean to add this file yet?

kiaraarose replied:
I did not! I'll remove it.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1213,6 +1215,73 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    // FIXME: <https://webkit.org/b/246483> Don't unload the background page if there are open ports, pending website requests, or if an inspector window is open.
+
+    if (m_backgroundWebView.get()._isBeingInspected) {
+        RELEASE_LOG(Extensions, "Not unloading non-persistent background content because it is being inspected.");
```
Should these logs contain an identifier for the extension? If so, should it be the context identifier or the unique identifier?

xeenon replied:
I think it should. The unique identifier.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1213,6 +1215,73 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    // FIXME: <https://webkit.org/b/246483> Don't unload the background page if there are open ports, pending website requests, or if an inspector window is open.
+
+    if (m_backgroundWebView.get()._isBeingInspected) {
+        RELEASE_LOG(Extensions, "Not unloading non-persistent background content for extension %{private}@ because it is being inspected.", static_cast<NSString*>(m_uniqueIdentifier));
```
You need to put a `%@` in the log to log the identifier. Also add a space in `static_cast<NSString *>`

xeenon replied:
Ooops, missed the `%{private}@`.

kiaraarose committed:
Queue and fire API events for Web Extensions.
https://bugs.webkit.org/show_bug.cgi?id=249254

Reviewed by Timothy Hatcher.

This patch adds support for queuing and firing API events for web extensions.

Testing:
I added an event listener in the background page of a test extension, and with this code below,
I tested that the completion handler was successfully called.
EventListenerTypeSet set;
set.add(WebExtensionEventListenerType::ActionOnClicked);
fireEvents(set, makeBlockPtr([ protectedThis = Ref { *this }]() {
    RELEASE_LOG(Extensions, "KIARA");
}));

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::loadBackgroundContentIfNecessary):
(WebKit::WebExtensionContext::scheduleBackgroundContentToUnload):

(WebKit::WebExtensionContext::queueStartupAndInstallEventsForExtensionIfNecessary):
Queue and fire event listeners.

(WebKit::WebExtensionContext::performTasksAfterBackgroundContentLoads):
(WebKit::WebExtensionContext::fireEvents):
Call the completion handler for persistent background pages. For non-persistent backgrounds, wake
up the background page and queue the listener to fire after the background is loaded.

(WebKit::WebExtensionContext::queueEventToFireAfterBackgroundContentLoads):
Add completion handler to queue of listeners to fires.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
(WebKit::WebExtensionAPIEvent::addListener):
(WebKit::WebExtensionAPIEvent::removeListener):
Add the extension context identifier. Without this, we won't know which context this message is for.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm:
(WebKit::WebExtensionAPIWebNavigationEvent::addListener):
(WebKit::WebExtensionAPIWebNavigationEvent::removeListener):
Add the extension context identifier. Without this, we won't know which context this message is for.

Canonical link: https://commits.webkit.org/257958@main
